modules/utask/execution.go

```diff
@@ -0,0 +1,152 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package utask
+
+import (
+	"encoding/gob"
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+// fnRegister allows looking up a function by name
+type fnRegister struct {
+	fnNameMap map[string]interface{}
+	sync.RWMutex
+}
+
+var fnLookup = fnRegister{fnNameMap: make(map[string]interface{})}
+
+// Signature represents a function and its arguments
+type Signature struct {
+	FnName string
+	Args   []interface{}
+}
+
+// Execute executes the function
+func (s *Signature) Execute() ([]reflect.Value, error) {
+	var targetArgs []reflect.Value
+	for _, arg := range s.Args {
+		targetArgs = append(targetArgs, reflect.ValueOf(arg))
+	}
+	fnLookup.RLock()
+	defer fnLookup.RUnlock()
+	if fn, ok := fnLookup.fnNameMap[s.FnName]; ok {
+		fnValue := reflect.ValueOf(fn)
+		return fnValue.Call(targetArgs), nil
+	}
+	return nil, fmt.Errorf("Function: %s not found. Did you forget to register?", s.FnName)
+}
+
+// Register registers a function for async tasks
+func Register(fn interface{}) error {
+	fnType := reflect.TypeOf(fn)
+	if err := validateAsyncFn(fnType); err != nil {
+		return err
+	}
+	for i := 0; i < fnType.NumIn(); i++ {
+		gob.Register(reflect.Zero(fnType.In(i)).Interface())
+	}
+	fnName := getFunctionName(fn)
+	fnLookup.Lock()
+	defer fnLookup.Unlock()
+	fnLookup.fnNameMap[fnName] = fn
+	return nil
+}
+
+// Enqueue sends a func before sending to the task queue
+func Enqueue(fn interface{}, args ...interface{}) error {
+	fnType := reflect.TypeOf(fn)
+	if fnType.NumIn() != len(args) {
+		return fmt.Errorf("Expected %d function arg(s) but found %d\n", fnType.NumIn(), len(args))
+	}
+	var argValues []reflect.Value
+	for i := 0; i < fnType.NumIn(); i++ {
+		arg := reflect.ValueOf(args[i])
+		argType := reflect.TypeOf(args[i])
+		if !argType.AssignableTo(fnType.In(i)) {
+			// TODO(madhu): Is it useful to show the arg index or the arg value in the error msg?
+			return fmt.Errorf(
+				"Cannot assign function argument: %d from type: %s to type: %s\n",
+				i+1, argType, fnType.In(i),
+			)
+		}
+		argValues = append(argValues, arg)
+	}
+	fnName := getFunctionName(fn)
+	s := Signature{FnName: fnName, Args: args}
+
+	sBytes, err := GlobalBackend().Encoder().Marshal(s)
+	if err != nil {
+		return errors.Wrap(err, "unable to encode the function")
+	}
+	return GlobalBackend().Publish(sBytes, nil)
+}
+
+// Run decodes the message and executes as a task
+func Run(message []byte) error {
+	var s Signature
+	if err := GlobalBackend().Encoder().Unmarshal(message, &s); err != nil {
+		return errors.Wrap(err, "unable to decode the message")
+	}
+	retValues, err := s.Execute()
+	if err != nil {
+		return err
+	}
+	return castToError(retValues[0])
+}
+
+// validateAsyncFn verifies that the type is a function type that returns only an error
+func validateAsyncFn(fnType reflect.Type) error {
```

Comment:
Could you rename it to validateFnReturnType? We don'e validate that the function is actually async :)